copybookreader-plugins/src/main/java/co/cask/hydrator/plugin/batch/CopyBookSource.java

```diff
@@ -0,0 +1,291 @@
+  /* For the initial implementation only fixed length binary files will be accepted.
+   This will not handle complex nested structures of COBOL copybook
+   Also this will not handle Redefines or iterators in the structure*/
+  public static final String COPYBOOK_INPUTFORMAT_FILE_STRUCTURE = "copybook.inputformat.input.filestructure";
+  public static final String COPYBOOK_INPUTFORMAT_DATA_HDFS_PATH = "copybook.inputformat.data.hdfs.path";
+
+  private final CopybookSourceConfig config;
+
+  public CopybookSource(CopybookSourceConfig copybookConfig) {
+    this.config = copybookConfig;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validate();
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(config.parseSchema());
+
+  }
+
+  @VisibleForTesting
+  final CopybookSourceConfig getConfig() {
+    return config;
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws IOException {
+    Job job = JobUtils.createInstance();
+    Configuration conf = job.getConfiguration();
+    conf.set(COPYBOOK_INPUTFORMAT_CBL_CONTENTS, config.copybookContents);
+    if (!(config.fileStructure.isEmpty())) {
+      conf.set(COPYBOOK_INPUTFORMAT_FILE_STRUCTURE, config.fileStructure);
+    }
+    conf.set(COPYBOOK_INPUTFORMAT_DATA_HDFS_PATH, config.binaryFilePath);
+    //set the input file path for job
+    CopybookInputFormat.setInputPaths(job, config.binaryFilePath);
+    context.setInput(Input.of(config.referenceName, new SourceInputFormatProvider(CopybookInputFormat.class, conf)));
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Map<String, String>> input, Emitter<StructuredRecord> emitter)
+    throws Exception {
+    Map<String, String> value = input.getValue();
+    Schema outputSchema;
+    if (config.schema == null) {
+      outputSchema = getSchema(value.keySet());
+    } else {
+      outputSchema = config.parseSchema();
+    }
+    StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+    for (Schema.Field field : outputSchema.getFields()) {
+      String fieldName = field.getName();
+      if (value.containsKey(fieldName)) {
+        builder.set(fieldName, parseValue(field, value.get(fieldName)));
+      }
+    }
+    emitter.emit(builder.build());
+  }
+
+  /**
+   * Convert each field to data type as specified by user in the output schema
+   *
+   * @param field Schema.Field object defining the field name and properties
+   * @param value value of field to be converted to datatype specified by user
+   * @return value casted to the required data type
+   * @throws IOException
+   */
+  private Object parseValue(Schema.Field field, String value) throws IOException {
+    Schema.Type fieldType = field.getSchema().getType();
+    if (fieldType.equals(Schema.Type.UNION)) {
```

Comment:
though this method in general seems very broken.  We should not be turning objects into strings then parsing them back into the original objects. Just use the original objects.